qiskit/primitives/containers/bindings_array.py

```diff
@@ -0,0 +1,355 @@
+        BindingsArray(np.empty((10, 10, 100)))
+
+        # multiple arrays, where each last index is parameters. notice that it's smart enough to
+        # figure out that a missing last dimension corresponds to a single parameter.
+        BindingsArray(
+            [np.empty((10, 10, 100)), np.empty((10, 10)), np.empty((10, 10, 20), dtype=complex)],
+            {("c", "a"): np.empty((10, 10, 2)), "b": np.empty((10, 10))}
+        )
+    """
+    __slots__ = ("_vals", "_kwvals")
+
+    def __init__(
+        self,
+        vals: ArrayLike | Iterable[ArrayLike] | None = None,
+        kwvals: Mapping[ParameterLike, Iterable[ParameterValueType]] | ArrayLike | None = None,
+        shape: ShapeInput | None = None,
+    ):
+        """
+        The ``shape`` argument does not need to be provided whenever it can unambiguously
+        be inferred from the provided arrays. Ambiguity arises because an array provided to the
+        constructor might represent values for either a single parameter, with an implicit missing
+        last dimension of size ``1``, or for many parameters, where the size of the last dimension
+        is the number of parameters it is providing values to. This ambiguity can be broken in the
+        following common ways:
+
+            * Only a single array is provided to ``vals``, and no arrays to ``kwvals``, in which case
+              it is assumed that the last dimension is over many parameters.
+            * Multiple arrays are given whose shapes differ only in the last dimension size.
+            * Some array is given in ``kwvals`` where the key contains multiple
+              :class:`~.Parameter` s, whose length the last dimension of the array must therefore match.
+
+        Args:
+            vals: One or more arrays, where the last index of each corresponds to
+                distinct parameters. If their dtypes allow it, concatenating these
+                arrays over the last axis is equivalent to providing them separately.
+            kwvals: A mapping from one or more parameters to arrays of values to bind
+                them to, where the last axis is over parameters.
+            shape: The leading shape of every array in these bindings.
+
+        Raises:
+            ValueError: If all inputs are ``None``.
+            ValueError: If the shape cannot be automatically inferred from the arrays, or if there
+                is some inconsistency in the shape of the given arrays.
+        """
+        super().__init__()
+
+        if vals is None:
+            vals = []
+        if kwvals is None:
+            kwvals = {}
+
+        vals = [vals] if isinstance(vals, np.ndarray) else [np.array(v, copy=False) for v in vals]
+        # TODO str will be used for internal data (_kwvals) instead of Parameter.
+        # This requires https://github.com/Qiskit/qiskit/issues/7107
+        kwvals = {
+            (p,) if isinstance(p, Parameter) else tuple(p): np.array(val, copy=False)
```

Comment:
This is incorrect with the type hint of `ParameterLike` which also allows strings. It should be
```suggestion
            (p,) if isinstance(p, (Parameter, str)) else tuple(p): np.array(val, copy=False)
```
or `ParameterLike` should be changed to `Parameter` to remove strings

Reply:
Certainly the type hint is wrong at this point. However, the decision has been made to wait for the str bind and I will correct it when I get it in.